### PR TITLE
Add Active Message receiver callbacks

### PR DESCRIPTION
**Pull request by pentschev**
Adds a new pattern to receive Active Messages where the receiving endpoint registers a callback and the sender informs receiver of the callback to use, the receiver then executes that callback and does not require an explicit `ep->amRecv()` call.

Sample code:

```cpp
// Define AM receiver callback's owner and id for callback
ucxx::AmReceiverCallbackInfo receiverCallbackInfo("TestApp", 0);

// Define AM receiver callback and register with worker
std::shared_ptr<ucxx::Request> receivedRequest = nullptr;
auto callback = ucxx::AmReceiverCallbackType(
  [this, &receivedRequest](std::shared_ptr<ucxx::Request> req) {
    {
      receivedRequest = req;
    }
  });
worker->registerAmReceiverCallback(receiverCallbackInfo, callback);

// Submit and wait for transfers to complete
std::shared_ptr<ucxx::Request> sendRequest;
requests.push_back(ep->amSend(buffer, size, memoryType, receiverCallbackInfo));
while (!sendRequest->isCompleted())
  worker->progress()

while (receivedRequest == nullptr)
  worker->progress();
```

This PR only implements C++ parts, https://github.com/rapidsai/ucxx/issues/187 tracks the missing Python implementation.

### CODE REVIEW

**Comment by pentschev** [comment]
@mdemoret-nv FYI

File: cpp/src/request_am.cpp [hunk 1]
```diff
@@ -5,17 +5,97 @@
...
+    bool hasReceiverCallback{static_cast<bool>(receiverCallbackInfo)};
+    const size_t ownerSize = (receiverCallbackInfo) ? receiverCallbackInfo->owner.size() : 0;
+    const size_t amReceiverCallbackInfoSize =
+      (receiverCallbackInfo) ? sizeof(ownerSize) + ownerSize + sizeof(receiverCallbackInfo->id) : 0;
+    const size_t totalSize =
+      sizeof(memoryType) + sizeof(hasReceiverCallback) + amReceiverCallbackInfoSize;
+    std::string serialized(totalSize, 0);
+
+    memcpy(serialized.data() + offset, &memoryType, sizeof(memoryType));
+    offset += sizeof(memoryType);
```
Comment (wence-) [inline cpp/src/request_am.cpp:78]: Very minor polish, perhaps unnecessary:
```
auto encode = [&offset, &serialized](void const *data, size_t bytes) {
  memcpy(serialized.data() + offset, data, bytes);
  offset += bytes;
}
```
And then:
```
encode(&memoryType, sizeof(memoryType));
encode(&hasReceiverCallback, sizeof(hasReceiverCallback));
if (hasReceiveCallback) {
  encode(&ownerSize, sizeof(ownerSize));
  encode(receiverCallbackInfo->owner.c_str(), ownerSize);
  encode(&receiverCallbackInfo->id, sizeof(receiverCallbackInfo->id));
}
```

And something matching on the decode side.

File: cpp/src/request_am.cpp [hunk 2]
```diff
@@ -119,15 +199,34 @@ ucs_status_t RequestAm::recvCallback(void* arg,
   bool is_rndv = param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV;
 
   std::shared_ptr<Buffer> buf{nullptr};
-  auto allocatorType = *static_cast<const ucs_memory_type_t*>(header);
+  // auto amHeader         = AmHeader::deserialize(static_cast<const char*>(header));
+  auto amHeader =
+    AmHeader::deserialize(std::string(static_cast<const char*>(header), header_length));
```
Comment (wence-) [inline cpp/src/request_am.cpp:204]: This copies the header, *suggestion*:
```suggestion
    AmHeader::deserialize(std::string_view(static_cast<const char*>(header), header_length));
```
And accept a `std::string_view` in `deserialize`.

**Review by wence- — APPROVED** [review]
Minor, I think

File: cpp/src/internal/request_am.cpp [hunk 3]
```diff
@@ -30,6 +32,12 @@ void RecvAmMessage::callback(void* request, ucs_status_t status)
 {
   std::visit(data::dispatch{
                [this, request, status](data::AmReceive amReceive) {
+                 if (_receiverCallback) {
+                   _request->_callback = [this](ucs_status_t, std::shared_ptr<void>) {
+                     _receiverCallback(_request);
+                   };
+                 }
+
                  _request->callback(request, status);
```
[the remark this answers is not in the document]
Comment (pentschev) [reply cpp/src/internal/request_am.cpp]: Done in https://github.com/rapidsai/ucxx/pull/186/commits/154274463ee286adb5dc624da2f728465f5d038b

File: cpp/src/request_am.cpp (hunk 1 above)
Comment (pentschev) [reply cpp/src/request_am.cpp]: That's an excellent idea! Done in https://github.com/rapidsai/ucxx/pull/186/commits/1920ff1c6b0ae71599e86bb9303b3d64e46545dc .

File: cpp/src/request_am.cpp (hunk 2 above)
Comment (pentschev) [reply cpp/src/request_am.cpp]: Done in https://github.com/rapidsai/ucxx/pull/186/commits/4304378c4b2ea578d97985782843a1bc04a405e0 .

**Comment by pentschev** [comment]
Thanks @wence- for reviewing!

**Pull request merged**